### DISCUSSION TITLE
Common attributes

### DISCUSSION THREAD

**Discussion by shivchelwa** [open-telemetry/opentelemetry-java#2494] [Q&A]
Is there a way to set common attributes such as app-name, module-name etc. on all the spans that will be created in my library/application? I could not locate any such feature. Please let me know if there an API available.

Thanks

**Comment by jkwatson** [comment]
It sounds like you're looking for the `Resource`. You can set a Resource when configuring your SDK, and it can have any common attributes you might want, and also be supplemented from our resource auto-populators as well.

**Reply by shivchelwa** [reply]
Hi @jkwatson , thanks for the quick reply. I tried using `Resource`, but I do not see attributes attached to span. Here is the code snippet I used to create a tracer. Is there something missing?

`		Resource resource = Resource.create(Attributes.builder().put("attr1", "value1").put("attr2", "value2").put("attr3", "value3").put("attr4", "value4").build());

		SpanExporter spanExporter = JaegerGrpcSpanExporter.builder()
										.setEndpoint("localhost:14250")
										.setServiceName("my-app")
										.build();

		BatchSpanProcessor spanProcessor = BatchSpanProcessor.builder(spanExporter).build();

		SdkTracerProvider sdkTraceProvider = SdkTracerProvider.builder().setResource(resource).build();

		sdkTraceProvider.addSpanProcessor(spanProcessor);

		OpenTelemetrySdk openTelemetry = OpenTelemetrySdk.builder().setTracerProvider(sdkTraceProvider).build();

		GlobalOpenTelemetry.set(openTelemetry);

		TracerProvider traceProvider = GlobalOpenTelemetry.get().getTracerProvider();

		Tracer tracer = traceProvider.get("my-library");`

**Reply by anuraaga** [reply]
Hi @shivchelwa - when using the jaeger exporter, resource attributes are attached as process tags, not span attributes. Can you find them there?

**Reply by shivchelwa** [reply]
Hi @anuraaga, Ahh! I see them under process tags. 

I have another observation. In my app, I want to enable auto-instrumentation as well(need to keep it optional for end-users). When I start the application with auto-instrumentation javaagent, I do not see the attributes set using `Resource` that I could see under process tags when jaeger exporter is used with manual instrumentation.

**Reply by anuraaga** [reply]
Auto instrumentation currently does not provide a programmatic way of setting resource, it only has a property that can be used to enable a key-value list.

https://github.com/open-telemetry/opentelemetry-java-instrumentation#opentelemetry-resource

Is it enough for your use case?

**Reply by shivchelwa** [reply]
Actually, its not about setting resource attributes specifically for auto-instrumentation. In my use-case, I am setting resource attributes on SdkTracerProvider in my code and it works fine if do not use auto-instrumentation. The same application code with auto-instrumentation enabled, does not use resource attributes set on SdkTracerProvider. Please refer to the code snippet in this thread. Is auto-instrumentation overwriting SdkTracerProvider created in code or not merging the resources?

**Reply by jkwatson** [reply]
If you're using the agent, any usage of the SDK will be ignored. The agent provides its own copy of the SDK and does not use any that you might configure.

**Reply by shivchelwa** [reply]
@jkwatson, that explains the behavior. But, it restricts use case like this, where some end-users of an application may want to enable javaagent to get details from instrumented libraries(cassandra, http), others may choose to just see traces from the application. Is there a way to accomplish this? OR is there a way to find out if javaagent is enabled, based on which I could initialize SDK.

**Reply by anuraaga** [reply]
@shivchelwa Nice timing, we just added a point to the documentation on how to configure the agent to only respect manual (application) instrumentation. Does it work for you?

https://github.com/open-telemetry/opentelemetry-java-instrumentation/commit/f3334c39909fef75cfb74614120fa12498d1da4b

**Reply by shivchelwa** [reply]
@anuraaga, looks like the suggested properties totally shuts out auto-instrumentation. I was looking for something to provide me traces from both auto and manual. With manual I can generate and capture telemetry from my application and with auto I could reap the benefits of instrumented libraries(which are used in my application).

And I could do that by enabling auto-instrumentation(javaagent) and instrumenting my application using API. But, in cases where javaagent is not used to start application, the instrumentation added in my application is useless. To address that, I used SDK to create SdkTracerProvider. But then, I see `Resource` set in manual instrumentation is not used when javaagent is used.

**Reply by anuraaga** [reply]
@shivchelwa Ah so you want to be able to run your app without the agent in a way that works, transparently? Currently I think we only support the case of lean vs non-lean by still having the agent in both cases but disabling instrumentation in the lean version. The agent intentionally hides the SDK to eliminate chances of version conflicts with an application. But we have been getting requests for more programatic configuration of the agent's SDK and are tracking it in #2002, I think we need to solve it (by first coming up with an idea for a way to solve it) to support this use case.

**Reply by shivchelwa** [reply]
@anuraaga, thank you for the explanation.
